crates/stages/api/src/pipeline/mod.rs

```diff
@@ -572,14 +572,18 @@ impl<N: ProviderNodeTypes> Pipeline<N> {
                     // FIXME: When handling errors, we do not commit the database transaction. This
                     // leads to the Merkle stage not clearing its checkpoint, and restarting from an
                     // invalid place.
-                    let provider_rw = self.provider_factory.database_provider_rw()?;
-                    provider_rw.save_stage_checkpoint_progress(StageId::MerkleExecute, vec![])?;
-                    provider_rw.save_stage_checkpoint(
-                        StageId::MerkleExecute,
-                        prev_checkpoint.unwrap_or_default(),
-                    )?;
-
-                    UnifiedStorageWriter::commit(provider_rw)?;
+                    // Only reset MerkleExecute checkpoint if MerkleExecute itself failed
```

Comment:
Without this check a state root error in MerkleChangeSets would result in the merkle checkpoint being cleared.